### PR TITLE
Support legacy mode for yyyymmdd format [databricks]

### LINKED ISSUE #11505: [FEA] Support yyyymmdd format for GetTimestamp for LEGACY mode.
**Is your feature request related to a problem? Please describe.**
```
!Expression <GetTimestamp> gettimestamp(finish_date#154, yyyymmdd, TimestampType, Some(Asia/Shanghai), false) cannot run on GPU because LEGACY format 'yyyymmdd' is not supported on the GPU.
```

```
spark.sql.legacy.timeParserPolicy=LEGACY
spark.rapids.sql.incompatibleDateFormats.enabled=true
```

### PR DESCRIPTION
**Pull request by res-life**
closes #11505

Required by customer.

Similar to this [PR](https://github.com/NVIDIA/spark-rapids/pull/11449) which supports `yyyyMMdd`, this PR supports `yyyymmdd`

Note: `MM` is month while `mm` is minute

Note:  GPU and CPU are not 100% consistent for `yyyymmdd` format, GPU only supports 8 digit strings.

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
> is `yyyymmdd` an legal format string for date ?
> 
> `yyyy` represents the year, but `mm` for minutes ? Isn't weird that a date is in the format of `{year}{minute}{day}` ?

In general, it's weird. But it's a valid format.  
`{year}{minute}{day}` does specify year, minute and day; the missed parts like `{month}{hour}{second}` have default values. e.g.: month default value is 1.

**Comment by firestarman** [comment]
> > is `yyyymmdd` an legal format string for date ?
> > `yyyy` represents the year, but `mm` for minutes ? Isn't weird that a date is in the format of `{year}{minute}{day}` ?
> 
> In general, it's weird. But it's a valid format. `{year}{minute}{day}` does specify year, minute and day; the missed parts like `{month}{hour}{second}` have default values. e.g.: month default value is 1.

I am not 100% sure on this, and would like to wait for the premerge result and verification in customer runs before approving it.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/date_time_test.py [hunk 1]
```diff
@@ -459,10 +459,12 @@ def test_to_timestamp(parser_policy):
             .select(f.col("a"), f.to_timestamp(f.col("a"), "yyyy-MM-dd HH:mm:ss")),
         { "spark.sql.legacy.timeParserPolicy": parser_policy})
 
+# mm: minute; MM: month
 @pytest.mark.skipif(not is_supported_time_zone(), reason="not all time zones are supported now, refer to https://github.com/NVIDIA/spark-rapids/issues/6839, please update after all time zones are supported")
 @pytest.mark.parametrize("format", ['yyyyMMdd', 'yyyymmdd'], ids=idfn)
 def test_formats_for_legacy_mode(format):
-    gen = StringGen("[0-9]{3}[1-9](0[1-9]|1[0-2])(0[1-9]|[1-2][0-9])")
+    # this regexp excludes zero year, python does not like zero year
+    gen = StringGen("([0-9]{3}[1-9])([0-5][0-9])([0-3][0-9])")
```
Comment (firestarman) [inline integration_tests/src/main/python/date_time_test.py:467]: This makes me wonder what will happen if an illegal input is given. We may need to test this case.

**Comment by pxLi** [comment]
build

File: integration_tests/src/main/python/date_time_test.py (hunk 1 above)
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py]: Illegal input produces NULL.
Added a test case which generates random 8 digits strings(exclude year 0000)

**Comment by res-life** [comment]
build

**Review by firestarman — APPROVED** [review]

**Pull request merged**